File: golang-http-backend/controller/product-controller.go
package controller

import (
	"connection-to-mongo/project/db"
	"connection-to-mongo/project/types"
	"connection-to-mongo/project/utils"
	"context"
	"fmt"
	"net/http"
	"time"

	"github.com/gorilla/mux"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

var ProductsController = http.HandlerFunc(productsController)

func productsController(w http.ResponseWriter, r *http.Request) {
	if r.Method == "GET" {
		getProducts(w, r)
		return
	} else {
		createProduct(w, r)
		return
	}
}

func createProduct(w http.ResponseWriter, r *http.Request) {
	var productWithImage types.ProductWithImageUrl
	var product types.Product
	var err error

	_ = utils.ParseJSON(r, &productWithImage)

	params := mux.Vars(r)

	storeId, _ := primitive.ObjectIDFromHex(params["storeId"])

	connection := db.ConnectToDB()
	defer db.CloseDB(connection)

	isAuthrorized := utils.IsAuthorizedForStore(r, storeId)

	if !isAuthrorized {
		utils.ResponseError(w, http.StatusNotAcceptable, fmt.Errorf(" unauthorized"))
		return
	}

	product.StoreId = storeId
	product.CategoryId = productWithImage.CategoryId
	product.SizeId = productWithImage.SizeId
	product.ColorId = productWithImage.ColorId
	product.Name = productWithImage.Name
	product.Price = productWithImage.Price
	product.IsArchived = productWithImage.IsArchived
	product.IsFeatured = productWithImage.IsFeatured
	product.CreatedAt = time.Now()
	product.UpdatedAt = time.Now()
	product.ID = primitive.NewObjectID()

	_, err = db.Products.InsertOne(context.Background(), product)
	if err != nil {
		utils.ResponseError(w, http.StatusConflict, fmt.Errorf("invalid storeId"))
		return
	}

	for _, url := range productWithImage.ImageUrl {
		var image types.Image

		image.CreatedAt = time.Now()
		image.UpdatedAt = time.Now()
		image.ProductId = product.ID
		image.Url = url

		_, err = db.Images.InsertOne(context.Background(), image)
	}

	if err != nil {
		utils.ResponseError(w, http.StatusConflict, fmt.Errorf("error while inserting a billboard"))
		return
	}

	utils.ResponseJSON(w, http.StatusOK, product)
}

func getProducts(w http.ResponseWriter, r *http.Request) {
	params := mux.Vars(r)

	storeId, err := primitive.ObjectIDFromHex(params["storeId"])

	if err != nil {
		utils.ResponseError(w, http.StatusNotAcceptable, fmt.Errorf("invalid store_id"))
		return
	}

	connection := db.ConnectToDB()
	defer db.CloseDB(connection)

	var products []types.FullProduct
	products, err = db.GetProductsByValue(bson.M{"storeId": storeId})
	if err != nil {
		utils.ResponseError(w, http.StatusConflict, fmt.Errorf("error while fecthing a products"))
		return
	}

	utils.ResponseJSON(w, http.StatusAccepted, products)
}

var UniqueProductController = http.HandlerFunc(uniqueProductController)

func uniqueProductController(w http.ResponseWriter, r *http.Request) {
	switch r.Method {

	case "GET":
		getProduct(w, r)
		return

	case "PATCH":
		updateProduct(w, r)
		return

	case "DELETE":
		deleteProduct(w, r)
		return
	}
}

func getProduct(w http.ResponseWriter, r *http.Request) {
	params := mux.Vars(r)

	storeId, err := primitive.ObjectIDFromHex(params["storeId"])

	if err != nil {
		utils.ResponseError(w, http.StatusNotAcceptable, fmt.Errorf("invalid store_id"))
		return
	}

	productId, err := primitive.ObjectIDFromHex(params["productId"])

	if err != nil {
		utils.ResponseJSON(w, http.StatusAccepted, nil)
		return
	}

	connection := db.ConnectToDB()
	defer db.CloseDB(connection)

	var product types.FullProduct
	filter := bson.M{"_id": productId, "storeId": storeId}
	db.GetProduct(&product, filter)

	utils.ResponseJSON(w, http.StatusAccepted, product)
}

func updateProduct(w http.ResponseWriter, r *http.Request) {
	var productWithImage types.ProductWithImageUrl
	var product types.Product
	var err error

	_ = utils.ParseJSON(r, &productWithImage)

	params := mux.Vars(r)

	productId, err := primitive.ObjectIDFromHex(params["productId"])
	if err != nil {
		utils.ResponseError(w, http.StatusNotAcceptable, fmt.Errorf("invalid cateogry_id"))
		return
	}

	storeId, _ := primitive.ObjectIDFromHex(params["storeId"])

	connection := db.ConnectToDB()
	defer db.CloseDB(connection)

	isAuthrorized := utils.IsAuthorizedForStore(r, storeId)

	if !isAuthrorized {
		utils.ResponseError(w, http.StatusNotAcceptable, fmt.Errorf(" unauthorized"))
		return
	}

	product.StoreId = storeId
	product.CategoryId = productWithImage.CategoryId
	product.SizeId = productWithImage.SizeId
	product.ColorId = productWithImage.ColorId
	product.Name = productWithImage.Name
	product.Price = productWithImage.Price
	product.IsArchived = productWithImage.IsArchived
	product.IsFeatured = productWithImage.IsFeatured
	product.CreatedAt = productWithImage.CreatedAt
	product.UpdatedAt = time.Now()
	product.ID = productId

	filter := bson.M{"_id": product.ID, "storeId": storeId}

	update := bson.M{"$set": product}

	_, err = db.Images.DeleteMany(context.Background(), bson.M{"productId": productId})

	if err != nil {
		utils.ResponseError(w, http.StatusConflict, fmt.Errorf("invalid storeId"))
		return
	}

	_, err = db.Products.UpdateOne(context.Background(), filter, update)
	if err != nil {
		utils.ResponseError(w, http.StatusConflict, fmt.Errorf("invalid storeId"))
		return
	}

	for _, url := range productWithImage.ImageUrl {
		var image types.Image

		image.CreatedAt = time.Now()
		image.UpdatedAt = time.Now()
		image.ProductId = product.ID
		image.Url = url

		_, err = db.Images.InsertOne(context.Background(), image)
	}

	if err != nil {
		utils.ResponseError(w, http.StatusConflict, fmt.Errorf("error while inserting a billboard"))
		return
	}

	utils.ResponseJSON(w, http.StatusOK, product)
}

func deleteProduct(w http.ResponseWriter, r *http.Request) {
	params := mux.Vars(r)

	storeId, err := primitive.ObjectIDFromHex(params["storeId"])
	if err != nil {
		utils.ResponseError(w, http.StatusNotAcceptable, fmt.Errorf("invalid store_id"))
		return
	}
	connection := db.ConnectToDB()
	defer db.CloseDB(connection)

	isAuthrorized := utils.IsAuthorizedForStore(r, storeId)

	if !isAuthrorized {
		utils.ResponseError(w, http.StatusNotAcceptable, fmt.Errorf(" unauthorized"))
		return
	}

	productId, err := primitive.ObjectIDFromHex(params["productId"])

	if err != nil {
		utils.ResponseError(w, http.StatusNotAcceptable, fmt.Errorf("invalid store_id"))
		return
	}

	filter := bson.M{"_id": productId, "storeId": storeId}
	_, err = db.Products.DeleteOne(context.Background(), filter)

	if err != nil {
		utils.ResponseError(w, http.StatusNotAcceptable, fmt.Errorf("invalid store_id"))
		return
	}

	_, err = db.Images.DeleteMany(context.Background(), bson.M{"productId": productId})

	if err != nil {
		utils.ResponseError(w, http.StatusConflict, fmt.Errorf("error while fecthing a products"))
		return
	}
	utils.ResponseJSON(w, http.StatusAccepted, "DELETED SUCCESSFULLY")
}

File: vanilla-nodejs-backend/src/controllers/product-controller.ts
import { IncomingMessage, ServerResponse } from "http";
import { getParams, getRequestBody, resultHeader } from "../utils/utils";
import { DBDeleteRecord,DBUpdateRecord, DBgetRecord, DBgetRecords, DBcreateSize, DBcreateProduct, DBcreateImage } from "../db/db-operation";
import { HttpStatusCode } from "axios";
import { Product } from "../type";

export const productsController = async ( req:IncomingMessage,res:ServerResponse) => {

    switch (req.method){
        case "POST" : createProduct(req,res)
                        return;
        case "GET" : getProducts(req,res)
                        return;
        default :
            res.writeHead(HttpStatusCode.MethodNotAllowed)
            res.write("METHOD NOT ALLOWED")
            res.end()
    }
}

const createProduct = async (req:IncomingMessage,res:ServerResponse) => {
    getRequestBody(req,async(data)=>{
        try {
            const {storeId} = getParams(req,'/store/:storeId/products')
            const userId = (req as any).userId
            const store = await DBgetRecord("store","id = ? AND userId = ?",storeId,userId)
            if(!store){
                res.writeHead(HttpStatusCode.Unauthorized)
                res.end("UnAuthorized")
                return
            }
            const {name,price,isFeatured,isArchived,sizeId,colorId,imageUrl,categoryId} = data
            const id = await DBcreateProduct(name,price,storeId,categoryId,isFeatured,isArchived,sizeId,colorId) as string
            for (let image of imageUrl){
                await DBcreateImage(image,id);
            }
            res.writeHead(HttpStatusCode.Created);
            res.write("PRODUCT CREATED")
            res.end()
            return
        } catch (error) {
            console.log("CREATE PRODUCT CONTROLLER",error)
            res.writeHead(HttpStatusCode.BadRequest);
            res.end("SOMETHING WENT WRONG")
        }
        
    })
}

const getProducts = async (req:IncomingMessage,res:ServerResponse)=>{
    try {
        const {storeId} = getParams(req,'/store/:storeId/products')
        let fullProducts :any[]= []
        const products = await DBgetRecords('product','storeId=?',storeId) as Product[]
        for (let product of products){
            let category = await DBgetRecord('category','id = ?',product.categoryId)
            let color = await DBgetRecord('color','id = ?',product.colorId)
            let size = await DBgetRecord('size','id = ?',product.sizeId)
            fullProducts.push({...product,category,color,size})
        }
        res.writeHead(HttpStatusCode.Accepted);
        res.write(JSON.stringify(fullProducts))
        res.end()

    } catch (error) {
        console.log("GET PRODUCTS CONTROLLER",error)
        res.writeHead(HttpStatusCode.InternalServerError)
        res.end("Something went wrong")
    }
}

export const uniqueProductController = async (req:IncomingMessage,res:ServerResponse)=>{
    switch (req.method){
        case "PATCH" : updateProduct(req,res)
                        return;
        case "GET" : getProduct(req,res)
                        return;
        case "DELETE" : deleteProduct(req,res)
                        return;
        default :
            res.writeHead(HttpStatusCode.MethodNotAllowed)
            res.write("METHOD NOT ALLOWED")
            res.end()
            return;
    }
}

export const getProduct = async (req:IncomingMessage,res:ServerResponse)=>{
    try {
        const {storeId,productId} = getParams(req,'/store/:storeId/products/:productId')
        if(productId == 'new'){
            res.writeHead(HttpStatusCode.Ok);
            res.end()
            return;
        }
        const product = await DBgetRecord('product','id = ? AND storeId = ? ',productId,storeId) as Product
        const category = await DBgetRecord('category','id = ? AND storeId = ? ',product.categoryId,storeId)
        const color = await DBgetRecord('color','id = ? AND storeId = ? ',product.colorId,storeId)
        const size = await DBgetRecord('size','id = ? AND storeId = ? ',product.sizeId,storeId)
        const image = await DBgetRecords('image','productId = ? ',product.id)
        if(!product){
            res.writeHead(HttpStatusCode.NotFound)
            res.end("NO PRODUCT FOUND")
            return
        }
        res.writeHead(HttpStatusCode.Accepted);
        res.write(JSON.stringify({...product,category,color,size,image}))
        res.end()
        return;
        
    } catch (error) {
        console.log("GET PRODUCT CONTROLLER",error)
        res.writeHead(HttpStatusCode.InternalServerError)
        res.end("Something went wrong")
    }
}

export const updateProduct = async (req:IncomingMessage,res:ServerResponse)=>{
    getRequestBody(req,async ( data )=>{
        try {
            const {storeId,productId} = getParams(req,'/store/:storeId/products/:productId')
            const userId = (req as any).userId
            const store = await DBgetRecord("store","id = ? AND userId = ?",storeId,userId)
            if(!store){
                res.writeHead(HttpStatusCode.Unauthorized)
                res.end("UnAuthorized")
                return
            }
            const {name,price,isFeatured,isArchived,sizeId,colorId,imageUrl,categoryId} = data
            await DBDeleteRecord('image','WHERE productId = ?',productId)
            await DBUpdateRecord('product','SET name = ?,price=?,isFeatured=?,isArchived=?,sizeId=?,colorId=?,categoryId=? WHERE id = ? AND storeId = ?'
                                ,name,price,isFeatured,isArchived,sizeId,colorId,categoryId,productId,storeId)
            for (let image of imageUrl){
                await DBcreateImage(image,productId);
            }
            res.writeHead(HttpStatusCode.Accepted);
            res.write("PRODUCT UPDATED SUCCESSFULLY")
            res.end()
        } catch (error) {
            console.log("UPDATE PRODUCT CONTROLLER",error)
            res.writeHead(HttpStatusCode.InternalServerError)
        res.end("Something went wrong")
        }
    })
}

export const deleteProduct = async (req:IncomingMessage,res:ServerResponse)=>{
        try {
            const {storeId,productId} = getParams(req,'/store/:storeId/products/:productId')
            const userId = (req as any).userId
            const store = await DBgetRecord("store","id = ? AND userId = ?",storeId,userId)
            if(!store){
                res.writeHead(HttpStatusCode.Unauthorized)
                res.end("UnAuthorized")
                return
            }
            await DBDeleteRecord('image','WHERE productId = ?',productId)
            const requestHeader = await DBDeleteRecord('product',' WHERE id = ? AND storeId = ?',productId,storeId) as resultHeader
            if(requestHeader.affectedRows < 1){
                res.writeHead(HttpStatusCode.NotFound);
                res.write("PRODUCT NOT FOUND")
                res.end()
                return
            }
            res.writeHead(HttpStatusCode.Accepted);
            res.write("PRODUCT DELETED SUCCESSFULLY")
            res.end()
        } catch (error) {
            console.log("DELETE PRODUCT CONTROLLER",error)
            res.writeHead(HttpStatusCode.InternalServerError)
            res.end("Something went wrong")
        }
}


